### ISSUE TITLE
Collection: support cross-document references with independent documents

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/metanorma-standoc#746]
From @anermina in metanorma/mn-samples-bsi#198 :

> PAS 2030:
> * Some cross-references were missing because the original document referred to the non-existing sections/tables.
>   For example "see 5.6" refers to the section that exists in PAS 2035, but not PAS 2030.
>   Same applies to referencing 4.2.4 and 4.2.5. Anchors are removed and these are now just a plain text which should be changed.
> 
> PAS 2035:
> * Similar as above, section 10.9.1 doesn't exist.

This is a great discovery that we need to handle. The question is when we have an independent cross-document citations from Document A to Document B, and when both documents are now put into a collection, how do we link between them?

We have to assume that Document A and Document B are independent documents and their encoding should not be changed.

It is up to the collection configuration to indicate that these two documents link through a bibliographic item, and that that links pointing to each other should be collection-internal links.

This means we have to either explicitly state the connection between an item inside a collection to another item's bibliography, or implicitly match via publication identifiers.

This also means that the semantic XML content from both Document A and Document B are unchanged. It is only when they are composed in a collection that this cross-document reference redirection occurs.

## Method 1: explicitly state connections

```yaml
collection:
  components:
    - document-root: pas-2030.adoc
      identifier: pas2030
      bibliographic-item-mapping:
        - "PAS 2035:2019": pas2035
    - document-root: pas-2035.adoc
      identifier: pas2035
      bibliographic-item-mapping:
        - "PAS 2030:2019": pas2030
```

## Method 2: implicit publication identifier match

In pas2030.adoc
```
= Title
:doctype: PAS
:docnumber: 2030
:copyright-year: 2019

...
== Bibliography
* [[[pas2035,PAS 2035:2019]]]
```

In pas2035.adoc
```
:doctype: PAS
:docnumber: 2035
:copyright-year: 2019

...
== Bibliography
* [[[pas2030,PAS 2030:2019]]]
```

Then when inside a collection, the bibliographic items automatically match because the identifiers can be resolved in a collection.

_Originally posted by @ronaldtse in https://github.com/metanorma/mn-samples-bsi/issues/198#issuecomment-1435514368_

**Comment by opoudjis** [comment]
Method 2 was always the intent, and the proposal of Method 1 is ridiculous:

````
collection:
  components:
    - document-root: pas-2030.adoc
      identifier: pas2030
      bibliographic-item-mapping:
        - "PAS 2035:2019": pas2035
    - document-root: pas-2035.adoc
      identifier: pas2035
      bibliographic-item-mapping:
        - "PAS 2030:2019": pas2030
````

Congratulations, you've just reinvented the manifest. WHICH IS ALREADY THERE.

````
manifest:
  level: collection
  title: BSI PAS 2035 / PAS 2030 collection
  docref:
    - fileref: _site/documents/pas-2035/document.xml
      identifier: PAS 2035:2020
    - fileref: _site/documents/pas-2030/document.xml
      identifier: PAS 2030:2019
````

**Comment by opoudjis** [comment]
In fact, Method 2 is already implemented: docidentifier PAS 2030:2019 in the bibliography is recognised as appearing in the manifest, and bibliographic links to it are replaced with direct links to the document.

So I have to work out what exactly you're finding is not implemented...

**Comment by opoudjis** [comment]
... The references should be working fine. Putting  a couple in that I can find.

Of course, references are only going to work if they match versions of standards in the manifest: `<<PAS_2035_2019,clause=5.6>>`, not `<<PAS_2035,clause=5.6>>`

**Comment by opoudjis** [comment]
@anermina The only 10.9.1 in PAS 2035 is in biblref_11, which is not PAS 2030. I can't find which citation you are referring to.

So I'm testing with the following two crossreferences, which are bibliographic in isolated documents, but which I expect to hyperlink to the other HTML document in a collection:

sources/pas-2030/sections/annex-b.adoc: `Retrofit Coordinator (<<PAS_2035_2019,clause=4.2.4;and!clause=4.2.5)`

sources/pas-2030/sections/annex-e.adoc: `minor editorial changes (see <<PAS_2035_2019,clause=5.6>>).`

If these work out, then proceed and insert other  crossreferences following that pattern. If not, then I will debug until you can.

**Comment by opoudjis** [comment]
Ah. I had updated isodoc to put NSBP for space in all docidentifiers being cited; I have not done same for metanorma collections rebuilding them, so those citations of docidentifiers impacted by the collection are not being found.

**Comment by opoudjis** [comment]
Rather disruptive change: since citeas is being used to look up file references, all indexes of file references have to be pretty-printed consistently.

**Comment by opoudjis** [comment]
Generating hyperlinks in this document has been a clusterfuck of confusion, and several of you have partaken in this confusion. So let me set some expectations here.

Hyperlinks between documents were working before, and have been fixed to work again, to allow for the change of inserting non-breaking spaces in citations.

However, if you think any random instance of a PAS 2030 is going to be hyperlinked successfully, you are very very much mistaken, and this is the fundamental problem of what you are trying to do. You need to decide what in fact you are doing.

PAS 2035, PAS 2035:2019, and PAS 2035:2020 are **COMPLETELY DIFFERENT DOCUMENTS**

Let me say that again, because I don't think people are realising this. They are  **COMPLETELY DIFFERENT DOCUMENTS**. 

(In fact, PAS 2035 is not a document at all, it is an abstraction. That is why style guides do not permit clause references on undated documents. We do not know from one version of the document to the next whether that document will exist.)

The **ONLY** document that is linked to from PAS 2030:2019 is PAS 2035:**2019**

The **ONLY** document that is included in this collection is PAS 2035:**2020**

If you expect hyperlinks to PAS 2035:**2019** to resolve to PAS 2035:**2020**, you are confused.

The document is not resolving links from PAS 2030. **CORRECTLY**. Because the version of the document they are linking to is different from the version of the document published.

If you want those links to work, you have to (silently) update the links from PAS 2030 to point to PAS 2039:2020, *and* you have to ensure that those links exist.

FWIW, the HTML hyperlinks from PAS 2035:2020 back to PAS 2030:2019 *do* work (though there aren't that many of them, and none of them are clause-specific).

Purely as a proof of concept, I am making the two links from PAS 2030:2019 be to PAS 2035:2020. The only way that is going to work is by inserting PAS 2035:2020 as a hidden (and genuinely anachronistic) reference in the bibliography of PAS 2030:2019.

@ronaldtse if you request me to make an alias of PAS 2035:2020 to PAS 2035:2019, I will answer that I *can* do that, but I will refuse to, because what is being attempted here is genuinely confused, and just itching for a lawsuit.

If you want to play silly buggers with the citation integrity of documents, you're going to end up playing with the documents. I have no particular guilt here, particularly as the anchor citation is hidden to begin with.

I'm attempting a workaround, not even confident it will work...

**Comment by opoudjis** [comment]
I am trying workarounds like

````
* [[[PAS_2035_2019,hidden(BSI PAS 2035:2019)]]] span:docid.BSI[PAS 2035:2020]
````

(Overwrite the 2019 record with a 2020 identifier)

It's failing on relaton because of https://github.com/relaton/relaton-bsi/issues/21

Honestly, I think the expedient thing would be to call the document in the collection PAS 2035:2019; after all, it's only 2020 because it includes a 2020 corrigendum.

Because I can't patch this, I'm putting this on hold, but you need to think a lot more clearly than you have been doing about what years to assign to documents in collections.

**Comment by opoudjis** [comment]
https://github.com/relaton/relaton-bsi/issues/21 unblocked, I'll see if I can patch the document to have the "wrong" docid.

**Comment by opoudjis** [comment]
!!!! It worked.

OK, so, if you want cross-references to work in this document between documents, but you've labelled the document PAS 2035:2020 on the collection manifest and cover page, but as PAS 2035:2019 in the actual document, you can change the document identifier to match the coverpage by override.
